python/pyspark/sql/column.py

```diff
@@ -406,7 +406,13 @@ def substr(self, startPos, length):
         [Row(col=u'Ali'), Row(col=u'Bob')]
         """
         if type(startPos) != type(length):
-            raise TypeError("Can not mix the type")
+            raise TypeError(
+                "startPos and length must be the same type. "
+                "Got {startPos_t} and {length_t}, respectively."
```

Comment:
-> `startPos: {startPos_t}; length: {length_t}.`

BTW, why we capture the type checking here, instead of doing it in the actual Scala impl of `substr`?